COHHIO/ryha
# Add geospatial filter

# Overview
Changes were applied to implement the following filter logic:

1. Each user will select one or more funders (all selected by default). Developed in #123 
2. Once funders are selected, users will select one or more counties. The list of counties will depend on funders selected. Only counties that contain projects from funders selected will be shown (all selected by default).
3. Finally, the user will select one or more projects. This list will show only projects from selected funders and counties (all selected by default).

Any change in selected funders or counties will recalculate the projects shown in projects filter, selecting all by default again. In other words, if when I open the app I select some projects (ignoring funder and counties filters) and then I change the funder, I will have to reselect the projects.

Basically, "Funders" and "Counties" filters are going to be used to subset the list of available projects to select from. This way, there is no need to alter filtering logic in other places of the code.

We assumed the workflow: Funders / Counties / Project to select the desired projects.

# List of Changes
- New internal dataset (`CountyCodes`) was added. 
    - `CountyCodes` contains two columns: `geocode` and `county`. 
    - Geocodes where extracted from https://www.hud.gov/sites/dfiles/CPD/documents/CoC/FY2024-GeoCodes-Report.pdf
    - Geocodes can refer to either cities or counties. Each city was manually assigned to the corresponding county.
    - We considered using a `geocodes.feather` file, but I think the proposed approach is easier to maintain in the future, e.g. if new cities are added to the project.
    - Data was split in cities and counties because VS Code was behaving in a weird way when trying to create a `tibble` with so many rows. The console wouldn't process the object creation. I think it is indeed a VS Code problem because I could create the file in RStudio without any issue. Anyways, the approach makes it straightforward to understand that there are two types of geographic regions.
- New "County" filter added.
    - Even though `geocodes` refer to cities for some projects, we decided to use counties as the geographic region for the filter. The problem with allowing users to select either cities or counties (based on the `geocode` value for the project) is that it could confuse users: users would see a list of cities and counties, but selecting a county wouldn't mean that a city included in the county is going to be selected. For this reason, users would have need to know whether their project `geocode` was a city or a county. Trying to explain this in a tooltip was also going to be problematic. We think that the county approach will be more straightforward for users. Also, as far as we checked, most of the programs with city `geocodes` actually serve individuals in multiple cities in their affiliated county.
    - "County" filter has a dependency on "Funder" filter: only counties with projects funded by selected funders are shown as valid choices.
    - "County" filter affects "Project" filter: in "Project" filter only projects funded by selected funders and located in selected counties are shown as valid choices.
    - The project filtering is done by actual county name because a county can have multiple `geocodes` associated with it. The "use `geocode` in the server but show the county name in the UI" approach resulted in repeated counties in the UI (as they were listed with different geocodes) so it was discarded.
- Developed logic to disable/enable "County" and "Project" filters based on user selections.
    - As there is a dependency between "Funder", "County" and "Project" filters, I thought it was good to incorporate some disable/enable functionality in the app to improve UX.
    - In addition to disabling the inputs, a popover message is displayed. To make this work, I had to wrap inputs in a new `div`. For some reason, I wasn't able to make it work by triggering the popover on the actual dropdown box. Even though the approach is not ideal, I think we can use it.
- Modified project coc in `create_dm()`
    - New `county` column has been added (it comes from joining ProjectCoC with `CountyCodes`).
    - An `"Unknown"` category is defined for codes that don't have a match in `CountyCodes`.
    - Instead of saving the value in the database on file upload, the label assignment is done when `dm` is being created. This removes the burden on the user to reupload data when a new code is added to `CountyCodes`
    - I think it makes sense to keep the actual `geocode` in the database, as it will help identifying codes with missing county or city labels.
 
# Test
- Open the project in the dev container
- Recreate dev database by running postgres/utils/recreate_dev.R
- Launch the app and upload some file
- Reopen the app to see the data
- Check that there is a new input between "Funder" and "Project" called "County"
- Deselect all counties and check that "Project" is disabled and that a popup message is shown asking the user to select at least one county
- Deselect all funders and check that "County" is disabled and that a popup message is shown asking the user to select at least one funder
- Check that "Project" updates accordingly based on "Funder" and "County" selections

Closes #45
